[14.0][ADD] wms_connector

kevinkhao opened the pull request:
The aim of wms_connector, is to provide a generic process for synchronizing picking with an external warehouse software (This can be useful when you subcontract your warehouse management to a WMS supplier)

Work is in progress

kevinkhao committed:
[ADD] wms_connector

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+        "ir.filters",
+        default=lambda r: r.env.ref("wms_connector.default_empty_filter"),
+    )
+    wms_picking_ar_filter_id = fields.Many2one(
```
```suggestion
    wms_picking_in_filter_id = fields.Many2one(
```

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+        "ir.filters",
+        default=lambda r: r.env.ref("wms_connector.default_empty_filter"),
+    )
+    wms_picking_prp_filter_id = fields.Many2one(
```
```suggestion
    wms_picking_out_filter_id = fields.Many2one(
```

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+    )
+    wms_product_sync_ids = fields.One2many("product.product", "warehouse_id")
+
+    def _inverse_active_wms_sync(self):
```
Maybe we can also create default "ir.filters" ? (filtering respectively on the in / out picking type ?)

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+            if task:
+                task.active = True
+            else:
+                setattr(
```
Can be done simpler

And please always add a _prepare_method when doing a create (it's make it inheritable for other module)

```
self[task_field_name] = self.env["attachement.synchronize.task"].create(
    self._prepare_sync_task(...)
)
```

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+            if cron:
+                cron.active = True
+            else:
+                setattr(
```
same as before.
- use self[cron_field_name] instead of setattr
- add prepare method

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+
+    def _deactivate_crons_tasks(self):
+        for kind in TASKS:
+            setattr(self, TASK_FIELDNAME.format(kind[0]), False)
```
self[ ...] = false

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,113 @@
+
+    def refresh_wms_products(self):
+        for rec in self:
+            rec.wms_product_sync_ids.unlink()
```
do not drop everything, it will be costly and you lose the sync information
Just search all product then
- create the missing binding
- unlink the binding that do not match the list of product

sebastienbeau commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,93 @@
+    wms_export_date = fields.Date()
+    wms_export_attachment = fields.Many2one("attachment.queue", index=True)
+    wms_export_error = fields.Char()
+    file_creation_mode = fields.Selection(
```
file_creation_mode is not a field on the mixing (you do not want to fill this info manually on every records)

Just a property in the class that return the file_creation_mode


```
@property
def file_creation_mode(self):
    return per_record
```

Then in the product export just inherit the class to change the value

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

kevinkhao committed:
wip

sebastienbeau commented on wms_connector/views/stock_warehouse.xml:
```diff
@@ -0,0 +1,65 @@
+                            type="object"
+                            string="Refresh WMS products"
+                        />
+                        <field name="wms_product_sync_ids">
```
Add a smart button that open the view in a dedicated view. Having a O2M integrated in the view with a lot of line is not usable

kevinkhao committed:
wip

kevinkhao committed:
wip

sebastienbeau commented on wms_connector/models/ir_cron.py:
```diff
@@ -0,0 +1,17 @@
+from odoo import fields, models
+
+
+class IrCron(models.Model):
+    _inherit = "ir.cron"
+
+    warehouse_export_ids = fields.One2many("stock.warehouse", "wms_export_cron_id")
+    warehouse_import_confirm_reception_ids = fields.One2many(
+        "stock.warehouse", "wms_import_confirm_reception_cron_id"
+    )
+    warehouse_import_confirm_delivery_ids = fields.One2many(
+        "stock.warehouse", "wms_import_confirm_delivery_cron_id"
+    )
+    product_sync_ids = fields.One2many("wms.product.sync", "warehouse_id")
```
Useless

sebastienbeau commented on wms_connector/models/attachment_synchronize_task.py:
```diff
@@ -0,0 +1,26 @@
+class AttachmentSynchronizeTask(models.Model):
+    _inherit = "attachment.synchronize.task"
+
+    def scheduler_export(self, model, domain=False):
+        if not domain:
+            recs = self.env[model].search([])
+        else:
+            recs = self.env[model].search(domain)
+        if not recs:
+            return
+        attachments = recs.with_context(attachment_task=self).synchronize_export()
+        attachments.run()
+
```
TODO, this should not be here

sebastienbeau commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,105 @@
+    @property
+    def file_creation_mode(self):
+        return "per_record"
+
```
TODO put the logic of the scheduler here

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,210 @@
+    wms_import_confirm_delivery_task_id = fields.Many2one(
+        "attachment.synchronize.task", readonly=True
+    )
+    wms_export_cron_id = fields.Many2one("ir.cron", readonly=True)
```
split this cron in 3

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,210 @@
+        "fieldname_cron": "wms_export_cron_id",
+        "filetype": "export",
+        "name_fragment": "exports (products, awaiting receptions, preparation orders",
+        "code": "wh = env['stock.warehouse'].browse({0})\n"
+        "wh.{1}.scheduler_export("
+        '"wms.product.sync", wh._wms_domain_for("wms_product_sync")'
+        ")\n"
+        "wh.{1}.scheduler_export("
+        '"stock.picking", wh._wms_domain_for("pickings_in")'
+        ")\n"
+        "wh.{1}.scheduler_export("
+        '"stock.picking", wh._wms_domain_for("pickings_out")'
+        ")",
```
env["wms.product.sync"]._schedule_export(warehouse)

on the wms.product.sync

def _schedule_export(self, warehouse):
     warehouse._sync_wms_product()
     return super()._schedule_export(warehouse)





env["stock.picking"]._schedule_export(warehouse, domain=wh._wms_domain_for("pickings_in") )

env["stock.picking"]._schedule_export(warehouse, domain=wh._wms_domain_for("pickings_out"))

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,210 @@
+
+    def refresh_wms_products(self):
+        for rec in self:
+            prd_with_sync = self.wms_product_sync_ids.product_id
```
existing_prd

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,210 @@
+                self.env["wms.product.sync"].create(
+                    {"product_id": prd.id, "warehouse_id": rec.id}
+                )
+            self.env["wms.product.sync"].search(
```
(existing_prd - matching_prd).wms_sync_ids.filtered(lambda s: s.warehouse_id = rec).unlink()

sebastienbeau commented on wms_connector/models/stock_warehouse.py:
```diff
@@ -0,0 +1,210 @@
+                or []
+            )
+            to_create = prd_matching - prd_with_sync
+            for prd in to_create:
```
TODO create with list

sebastienbeau commented on wms_connector/data/cron.xml:
```diff
@@ -0,0 +1,19 @@
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo noupdate="1">
+
+    <record model="ir.cron" id="refresh_wms_products">
```
Remove this cron. Use the export_product_cron

sebastienbeau commented on wms_connector/data/ir_filters.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2023 Akretion
+     License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl). -->
+<odoo noupdate="1">
```
Remove data and create on the fly, when activating the sync on wms

sebastienbeau replied:
to filter on the picking_type. [("picking_type_id", "=", wh.out_type_id.id)]

sebastienbeau replied:
also on picking filter on the state ("state", "=", "assigned")

sebastienbeau commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,105 @@
+                        attachment = rec._format_to_exportfile(data)
+                        rec.track_export(attachment)
+                        res += attachment
+                    except Exception as e:
```
from odoo.tool import config

```
if config.get("dev") in ("pdb", "all"):
   raise
```

sebastienbeau commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,105 @@
+        self.synchronize_export()
+
+    def synchronize_export(self):
+        if self.file_creation_mode == "per_record":
```
```
def _get_export_data(self):
    data = []
    for record in self:
             try:
                  data.append(record._prepare_export_data())
             except:
                  record.error = ""
                  continue
        if record._file_creation_mode == "per_record":
            yield data
            data  = []
    if record._file_creation_mode == "per_recordset":
        yield data

for data in self._get_export_data():
    self._format_to_exportfile(data)

```

kevinkhao committed:
change cron and filter logic

kevinkhao committed:
change export logic

kevinkhao committed:
cleanup

kevinkhao committed:
misc fixes, tests

kevinkhao committed:
minor name fix

kevinkhao committed:
minor interface fix

sebastienbeau committed:
wms_connector: improve export

sebastienbeau committed:
wms_connector: clean code, improve view

Now you can define the number of record per file
binding are not deleted anymore in order to keep the history

sebastienbeau committed:
wms_connector: fix export, review product sync

bealdav commented on wms_connector/models/attachment_synchronize_task.py:
```diff
@@ -0,0 +1,16 @@
+
+    file_type = fields.Selection(
+        selection_add=[
+            ("export", "Export"),
```
nitpicking : `export` should be exist in `attachment_synchonize` module and not be here.
Can be solved later

sebastienbeau committed:
wms_connector: do not try to send record with error

FranzPoize committed:
wms_connector: ADD make filetype consistent between task and queue

FranzPoize committed:
wms_connector: ADD adds method_type and filepath to task mappings for default task creation

jbaudoux commented:
Shouldn't this go to edi repo ?

FranzPoize committed:
FIX method_type and filepath default values

FranzPoize committed:
ADD storage backend value (mostly for test purpose right now)

sebastienbeau committed:
wms_connector: improve code

- do not add storage_backend_id on warehouse, to avoid extra complexity when using server env
- add explicit method_type and filepath

bealdav commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,97 @@
+    _description = "Synchronizable export mixin"
+
+    wms_export_date = fields.Datetime(readonly=True)
+    wms_export_attachment = fields.Many2one(
```
broken convention
```suggestion
    wms_export_attachment_id = fields.Many2one(
```

FranzPoize committed:
wms_connector: ADD wms_connector_exported and is_wms_exportable fields

This allows to display the export to wms contextually and display wms_connector export state in tree view

FranzPoize committed:
wms_connector: ADD a smart button to the picking view form that links to
the wms export

sebastienbeau committed:
wms_connector: clean view, and simplify code, add cancel check

sebastienbeau committed:
Merge pull request #8 from akretion/14.0-wms_connector_3

14.0 wms connector 3

FranzPoize committed:
wms_connector: ADD inventory import support

FranzPoize committed:
wms_connector: compute default_warehouse_id for inventory in a simpler way

sebastienbeau committed:
wms_conne: return attachment queue

sebastienbeau committed:
wms_connector: fix getting task, show error message on picking, raise error when using button

sebastienbeau committed:
wms_connector: fix access right issue

sebastienbeau committed:
wms_connector: fix cancelling sale order. Do not raise error if already cancel

bealdav commented on wms_connector/models/synchronize_importable_mixin.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class SynchronizeImportableMixin(models.AbstractModel):
+    _name = "synchronize.importable.mixin"
```
I don't know where this mixin is make it concrete. It should be pickings ?

kevinkhao replied:
for public discourse: 

can be applied to pickings, inventories and whatever else

bealdav commented on wms_connector/models/synchronize_exportable_mixin.py:
```diff
@@ -0,0 +1,107 @@
+    wms_export_attachment_id = fields.Many2one(
+        "attachment.queue", index=True, readonly=True
+    )
+    wms_export_error = fields.Char(readonly=True, index=True)
```
This mixin is used in pickings, OK. A picking is either wms exportable or wms importable, isn't it ?

If yes fields names should be `wms_date`, `wms_attachment_id`, `wms_error` ou wms_exception.

Qu'en pensez vous ?

bealdav commented on wms_connector/models/attachment_queue.py:
```diff
@@ -0,0 +1,27 @@
+
+WMS_IMPORT_FILETYPES = [
+    ("wms_reception_confirmed", "WMS Reception confirmed"),
+    ("wms_delivery_confirmed", "WMS Delivery confirmed"),
```
```suggestion
    ("wms_delivery_confirmed", "WMS Delivery confirmed"),
    ("wms_inventory", "WMS Inventory"),
```

bealdav commented on wms_connector/models/attachment_queue.py:
```diff
@@ -0,0 +1,27 @@
+        raise NotImplementedError
+
+    def _run_wms_delivery_confirmed(self):
+        raise NotImplementedError
```
```suggestion
        raise NotImplementedError
        
    def _run_wms_inventory(self):
        raise NotImplementedError

```

sebastienbeau committed:
Merge remote-tracking branch 'origin/14.0-wms_connector_inventory_support' into 14.0-wms_connector

sebastienbeau committed:
wms_connector: improve UI, clean code

sebastienbeau committed:
wms_connector: fix computing warehouse_id

sebastienbeau committed:
wms_connector: purge binding that have been never exported, and clean code

sebastienbeau commented:
@jbaudoux  indeed maybe it will be better to have it in edi repo
@bealdav what do you think ?

sebastienbeau committed:
wms_connector: reset wms_export_date and file when force cancel, so we can re-export

sebastienbeau committed:
wms_connector: do not unlink file sent

sebastienbeau committed:
wms_connector: fix checking allowed to cancel

sebastienbeau committed:
wms_connector: do not edit picking already synchronized. Add fr translation

FranzPoize committed:
wms_connector: Allow sequence management by child specialization

FranzPoize committed:
wms_connector: encapsulate MAPPINGS and FILTER VAlS in function for inheritance

FranzPoize committed:
wms_connector_picking_batch: Adds picking batch specialization for wms_connector

FranzPoize committed:
wms_connector_picking_batch: Adds README.rst

FranzPoize committed:
wms_connector_picking_batch: good manifest

FranzPoize committed:
wms_connector_picking_batch: working filters

kevinkhao commented on wms_connector_picking_batch/models/stock_picking_batch.py:
```diff
@@ -0,0 +1,73 @@
+from odoo.exceptions import UserError
+
+
+class StockPikcingBatch(models.Model):
```
minor typo

kevinkhao commented:
https://github.com/OCA/server-tools/pull/2883
might be of interest to avoid repeating some logic

FranzPoize committed:
wms_connector: allows to specify record_per_file to -1 which puts all the record in one file

FranzPoize committed:
wms_cconnector_picking_batch: fix wms_export_task_id access

FranzPoize committed:
wms_connector: makes filter_domains easier to modify by inheriting models

FranzPoize committed:
wms_connector_picking_batch: modify filter_domains

FranzPoize committed:
wms_connector_picking_batch: fix stock_picking modal in stock_picking_batch

FranzPoize committed:
wms_connector: adds missing field to stock_picking_views

FranzPoize committed:
wms_connector_picking_batch: adds field to stock_picking_batch

FranzPoize committed:
wms_connector: remove incoming and outgoing constraints on exportable picking

FranzPoize committed:
wms_connector_picking_batch: Choose batch transfer in state draft

FranzPoize committed:
[FIX] make it possible to export errored wms_product_sync without changing its name
